### ISSUE TITLE
Build JSON pointers map for OpenApi/Swagger objects

### ISSUE THREAD

**Issue by vadeg** [zalando/zally#689]
Right now we need to manually build path for each property for every rule. We want to do it automatically. For this we need to build JSON pointers map out of `OpenApi`/`Swagger` objects. Then we can get property pointers from the map instead of building them by their own.

**Comment by roxspring** [comment]
Intriguing.

Is the expectation here that Violation.paths will be switched from `List<String>` to `List<JsonPath>`?
Will the path lookup be happening in the RulesValidator, or will you be passing the map down to the rules for them to make use of?

**Comment by vadeg** [comment]
@roxspring 

> Is the expectation here that Violation.paths will be switched from List<String> to List<JsonPath>?
We will keep it as is.

> Will the path lookup be happening in the RulesValidator, or will you be passing the map down to the rules for them to make use of?

Map will be build before rules validation starts. Lookup will be done inside rule. Basically it is up to rule author when to lookup json pointer for `OpenApi`/`Swagger` object.

**Comment by roxspring** [comment]
So rule implementations would go something like the following? (with whatever the valid json-pointer expresssion would be)
```
@Check(Severity.MUST)
fun validate(context: Context): Violation? =
    when (context.resolveJsonPath("/paths/[/things]")) {
        is a Path -> // validate it as Swagger Path
        is a PathItem -> // validate it as OpenAPI PathItem
        else -> null
```
I definitely welcome the experiment but I'm not seeing much benefit for the rule developer compared with other approaches.

**Comment by mfellner** [comment]
No, actually what we mean is the other way around.

Basically, right now the "paths" for the generated violation (`Violation.paths: List<String>`) have to be constructed manually with string concatenation and other convulated ways in every rule, e.g. `"#/definitions/$defName/properties/$propName" + something` (ExtensibleEnumRule.kt).

In other words, *accessing and evaluating* some child node(s) in `Swagger` and *violation path construction* is entangled. Additionally, **not all** violation paths are actually valid [JSON pointers](https://tools.ietf.org/html/rfc6901) (e.g. `SnakeCaseForQueryParamsRule`).

So what we want instead is to be able to access some child node(s) in `Swagger`, evaluate it, and then get a valid JSON pointer string to that node. This can be achieved by having a lookup `Map` for every unique child node instance (identified by `identityHashCode`) to the JSON pointer string of that node.

This `Map<Integer, String>` would be created ahead of time by traversing `Swagger` recursively using reflection.

For example:

```kotlin
val paths = swagger.paths.flatMap { path ->
    path.value.operations.orEmpty().flatMap { operation ->
        if (checkIsViolation(operation)) {
            identityToPath[System.identityHashCode(operation)]?.let {
                listOf(it)
            } ?: emptyList()
        } else {
            emptyList()
        }
    }
}
```

The `Map` can live inside the `Context` introduced in https://github.com/zalando/zally/pull/668. Plus, traversing the `Swagger` object ahead of time can be used for additional "cleanup" before running the checks (e.g. ignoring/filtering sub-trees of `Swagger`).

**Comment by roxspring** [comment]
> In other words, accessing and evaluating some child node(s) in Swagger and violation path construction is entangled. 

Agree. Dis-entangling would be good.

> Additionally, not all violation paths are actually valid JSON pointers (e.g. SnakeCaseForQueryParamsRule).

Agree again. In-house we're regularly abusing the Violation.paths to contain both the path AND some path specific error message.

> So what we want instead is to be able to access some child node(s) in Swagger, evaluate it, and then get a valid JSON pointer string to that node. This can be achieved by having a lookup Map for every unique child node instance (identified by identityHashCode) to the JSON pointer string of that node.

... and that JSON pointer string would be used in the violation somewhere, possibly as part of the Violation.paths, but without enforcing that at any type level for now?

Sounds like some good steps forward.

**Comment by mfellner** [comment]
However it means rewriting all the rules to make use of the lookup map. On the other hand that change would be backwards compatible, not all rules must make use of the map and can be left unchanged.

Then again creating the list of paths for a `Violation` already puts a lot of burden on a rule implementation. I could imagine changing the `Violation` class to have `description: String` but `nodes: List<Object>` instead of `paths: List<String>`. Then a rule would just return the list of *violating nodes* and it would be the job of `RulesValidator` to unpack that list into a list of JSON pointer paths or do something else with it.

If you abuse `paths` to store additional information, maybe we should think about extending `Violation` to cover your use cases? Or instead, maybe a rule should return a *list* of violations which each violation having **1** `node: Object` and **1** `description: String`?

**Comment by roxspring** [comment]
All that sounds good.

I wouldn't worry too much about breaking backwards compatibility (at Java/Kotlin level) with existing rules - I was certainly encouraged not to when switching to using annotations. I do think Java/Kotlin API needs to freeze at some point but would rather get it right first.

One nice aspect of returning a list of (node,description) pairs is that it would also give us the option of grouping and displaying results by location rather than severity & rule. Eventually, when we can also map a JSON Pointer to a line and column number, I can automatically comment on the offending lines in GitLab merge requests!!

**Comment by mkulak** [comment]
My 5 cents
Rather than this `identityToPath[System.identityHashCode(operation)]` consider using [IdentityHashMap](https://docs.oracle.com/javase/7/docs/api/java/util/IdentityHashMap.html)
It is more concise and will handle hash collisions gracefully.

**Comment by mfellner** [comment]
@mkulak thanks, you're right. We ended up using IdentityHashMap too! https://github.com/zalando/zally/blob/openapi3/server/src/main/java/de/zalando/zally/util/OpenApiWalker.kt

**Comment by roxspring** [comment]
Pretty sure this is all done now. Shall we close the issue @maxim-tschumak?

**Comment by maxim-tschumak** [comment]
Implemented in #706
